File: app/src/main/java/com/example/myapplication/protocol/SensorBlockParser.kt
package com.example.myapplication.protocol

data class SensorBlock(
    val sensorType: Int,
    val channelSamples: List<List<Float>>,
)

data class SensorBlockParseResult(
    val blocks: List<SensorBlock>,
    val parsedBlockCount: Int,
    val consumedBytes: Int,
    val isComplete: Boolean,
)

class SensorBlockParser {
    fun parse(packetBytes: ByteArray): List<SensorBlock> {
        return parseResult(packetBytes).blocks
    }

    fun parseResult(packetBytes: ByteArray): SensorBlockParseResult {
        if (packetBytes.size < HEADER_SIZE) {
            return SensorBlockParseResult(
                blocks = emptyList(),
                parsedBlockCount = 0,
                consumedBytes = 0,
                isComplete = false,
            )
        }

        val measurementCount = packetBytes[6].toInt() and 0xFF
        var blockOffset = HEADER_SIZE
        val blocks = mutableListOf<SensorBlock>()
        var parsedBlockCount = 0

        repeat(measurementCount) {
            if (blockOffset + BLOCK_HEADER_SIZE > packetBytes.size) {
                return SensorBlockParseResult(
                    blocks = blocks.toList(),
                    parsedBlockCount = parsedBlockCount,
                    consumedBytes = blockOffset,
                    isComplete = false,
                )
            }

            val blockSensorType = packetBytes[blockOffset].toInt() and 0xFF
            val channelCount = packetBytes[blockOffset + 1].toInt() and 0xFF
            val samplesPerChannel = (packetBytes[blockOffset + 2].toInt() and 0xFF) or
                ((packetBytes[blockOffset + 3].toInt() and 0xFF) shl 8)

            val blockPayloadSize = channelCount * samplesPerChannel * BYTES_PER_SAMPLE
            val nextBlockOffset = blockOffset + BLOCK_HEADER_SIZE + blockPayloadSize
            if (nextBlockOffset > packetBytes.size) {
                return SensorBlockParseResult(
                    blocks = blocks.toList(),
                    parsedBlockCount = parsedBlockCount,
                    consumedBytes = blockOffset,
                    isComplete = false,
                )
            }

            val channels = MutableList(channelCount) { mutableListOf<Float>() }
            var sampleOffset = blockOffset + BLOCK_HEADER_SIZE

            for (channelIndex in 0 until channelCount) {
                repeat(samplesPerChannel) {
                    val raw = (packetBytes[sampleOffset].toInt() and 0xFF) or
                        ((packetBytes[sampleOffset + 1].toInt() and 0xFF) shl 8)
                    channels[channelIndex] += raw.toShort().toFloat()
                    sampleOffset += BYTES_PER_SAMPLE
                }
            }

            blocks += SensorBlock(
                sensorType = blockSensorType,
                channelSamples = channels.map { it.toList() },
            )
            blockOffset = nextBlockOffset
            parsedBlockCount++
        }

        return SensorBlockParseResult(
            blocks = blocks.toList(),
            parsedBlockCount = parsedBlockCount,
            consumedBytes = blockOffset,
            isComplete = blockOffset == packetBytes.size,
        )
    }

    fun extractChannelSamples(
        packetBytes: ByteArray,
        sensorType: Int,
        channel: Int,
    ): List<Float> = extractChannelSamples(
        sensorBlocks = parse(packetBytes),
        sensorType = sensorType,
        channel = channel,
    )

    fun extractChannelSamples(
        sensorBlocks: List<SensorBlock>,
        sensorType: Int,
        channel: Int,
    ): List<Float> {
        if (channel <= 0) return emptyList()

        return sensorBlocks
            .asSequence()
            .filter { it.sensorType == sensorType && channel <= it.channelSamples.size }
            .flatMap { it.channelSamples[channel - 1].asSequence() }
            .toList()
    }

    private companion object {
        const val HEADER_SIZE = 16
        const val BLOCK_HEADER_SIZE = 6
        const val BYTES_PER_SAMPLE = 2
    }
}

File: ios/Woona/Protocol/SensorBlockParser.swift
import Foundation

struct SensorBlock: Equatable {
    let sensorType: Int
    let channelSamples: [[Float]]
}

struct SensorBlockParseResult: Equatable {
    let blocks: [SensorBlock]
    let parsedBlockCount: Int
    let consumedBytes: Int
    let isComplete: Bool
}

struct SensorBlockParser {
    func parse(_ packetBytes: [UInt8]) -> [SensorBlock] {
        parseResult(packetBytes).blocks
    }

    func parseResult(_ packetBytes: [UInt8]) -> SensorBlockParseResult {
        guard packetBytes.count >= Self.headerSize else {
            return SensorBlockParseResult(
                blocks: [],
                parsedBlockCount: 0,
                consumedBytes: 0,
                isComplete: false
            )
        }

        let measurementCount = Int(packetBytes[6])
        var blockOffset = Self.headerSize
        var blocks: [SensorBlock] = []
        var parsedBlockCount = 0

        for _ in 0..<measurementCount {
            guard blockOffset + Self.blockHeaderSize <= packetBytes.count else {
                return SensorBlockParseResult(
                    blocks: blocks,
                    parsedBlockCount: parsedBlockCount,
                    consumedBytes: blockOffset,
                    isComplete: false
                )
            }

            let sensorType = Int(packetBytes[blockOffset])
            let channelCount = Int(packetBytes[blockOffset + 1])
            let samplesPerChannel = Int(packetBytes[blockOffset + 2]) |
                (Int(packetBytes[blockOffset + 3]) << 8)
            let payloadSize = channelCount * samplesPerChannel * Self.bytesPerSample
            let nextBlockOffset = blockOffset + Self.blockHeaderSize + payloadSize

            guard nextBlockOffset <= packetBytes.count else {
                return SensorBlockParseResult(
                    blocks: blocks,
                    parsedBlockCount: parsedBlockCount,
                    consumedBytes: blockOffset,
                    isComplete: false
                )
            }

            var channels = Array(repeating: [Float](), count: channelCount)
            var sampleOffset = blockOffset + Self.blockHeaderSize

            for channelIndex in 0..<channelCount {
                for _ in 0..<samplesPerChannel {
                    let raw = UInt16(packetBytes[sampleOffset]) |
                        (UInt16(packetBytes[sampleOffset + 1]) << 8)
                    channels[channelIndex].append(Float(Int16(bitPattern: raw)))
                    sampleOffset += Self.bytesPerSample
                }
            }

            blocks.append(SensorBlock(sensorType: sensorType, channelSamples: channels))
            blockOffset = nextBlockOffset
            parsedBlockCount += 1
        }

        return SensorBlockParseResult(
            blocks: blocks,
            parsedBlockCount: parsedBlockCount,
            consumedBytes: blockOffset,
            isComplete: blockOffset == packetBytes.count
        )
    }

    func extractChannelSamples(
        packetBytes: [UInt8],
        sensorType: Int,
        channel: Int
    ) -> [Float] {
        extractChannelSamples(sensorBlocks: parse(packetBytes), sensorType: sensorType, channel: channel)
    }

    func extractChannelSamples(
        sensorBlocks: [SensorBlock],
        sensorType: Int,
        channel: Int
    ) -> [Float] {
        guard channel > 0 else { return [] }

        return sensorBlocks.flatMap { block -> [Float] in
            guard block.sensorType == sensorType, channel <= block.channelSamples.count else {
                return []
            }
            return block.channelSamples[channel - 1]
        }
    }

    private static let headerSize = 16
    private static let blockHeaderSize = 6
    private static let bytesPerSample = 2
}
